cai_causal_graph/graph_components.py

```diff
@@ -666,6 +584,66 @@ def to_dict(self, include_meta: bool = True) -> dict:
         }
 
         if include_meta:
-            edge_dict['meta'] = self.meta
+            edge_dict['meta'] = self.meta.copy()
 
         return edge_dict
+
+
+class TimeSeriesEdge(Edge):
+    """
+    Class defining a time-series edge.
+
+    Time series edge is equivalent to a `cai_causal_graph.graph_components.Edge` class, except that it only
+    ensures that the destination of an edge is at the same or later time than its source.
+
+    This means that it is not possible to construct a directed time-series edge which does not respect time. Moreover,
+    if an undirected edge is constructed by passing a source which is at a later time than the destination, the source
+    and destination are swapped.
+    """
+
+    _NodeClassDict = Edge._NodeClassDict.copy()
+    _NodeClassDict.update({TimeSeriesNode.__name__: TimeSeriesNode})
+
+    source: TimeSeriesNode
+    destination: TimeSeriesNode
+
+    def __init__(
+        self,
+        source: Node,
+        destination: Node,
+        edge_type: EdgeType = EdgeType.DIRECTED_EDGE,
+        meta: Optional[Dict[str, Any]] = None,
+    ):
+        """
+        Construct a `TimeSeriesEdge`.
+
+        :param source: The `cai_causal_graph.graph_components.Node` from which the edge will originate.
+        :param destination: The `cai_causal_graph.graph_components.Node` at which the edge will terminate.
+        :param edge_type: The type of the edge to be added. Default is
+            `cai_causal_graph.type_definitions.EdgeType.DIRECTED_EDGE`. See `cai_causal_graph.type_definitions.EdgeType`
+            for the list of possible edge types.
+        :param meta: The meta values for the edge. Default is `None`. If passed, meta is shallow-copied.
+        """
+        if not isinstance(source, TimeSeriesNode):
+            source = TimeSeriesNode.from_dict(source.to_dict(include_meta=True))
+
+        assert isinstance(source, TimeSeriesNode)   # for lint
+
+        if not isinstance(destination, TimeSeriesNode):
+            destination = TimeSeriesNode.from_dict(destination.to_dict(include_meta=True))
+
+        assert isinstance(destination, TimeSeriesNode)   # for lint
+
+        # If edge type is not directed, swap source and destination to respect time
+        if edge_type != EdgeType.DIRECTED_EDGE:
+            if source.time_lag > destination.time_lag:
+                _destination = source
+                source = destination
+                destination = _destination
```

Comment:
Completely unnecessary optimization but I discovered recently that you can do this to swap variables in one line 🙃
```
source, destination = destination, source
```